### ISSUE TITLE
`default` and `default_factory` do not work in nested list with parse_object

### ISSUE THREAD

**Issue by function2-llx** [omni-us/jsonargparse#357]
### To reproduce

```python
from dataclasses import dataclass, field

from jsonargparse import ArgumentParser

@dataclass
class A:
    x: list = field(default_factory=list)
	v: int = 1

@dataclass
class B:
    a: list[A]

def main():
    parser = ArgumentParser()
    parser.add_dataclass_arguments(B, 'conf')
    args = parser.parse_object({
        'conf': {
            'a': [{}]
        }
    })
    print(args.conf)

if __name__ == '__main__':
    main()
```

### Expected result

```
Namespace(a=[Namespace(v=1, x=[])])
```

### Actual result

```
Namespace(a=[Namespace()])
```

### Environment

- jsonargparse version (e.g., 4.8.0): 4.23.1
- Python version (e.g., 3.9): 3.11.4
- How jsonargparse was installed: `pip install jsonargparse[all]`
- OS (e.g., Linux): Ubuntu 22.04

**Comment by mauvilsa** [comment]
Thank you for reporting!

**Comment by mauvilsa** [comment]
I implemented a fix in #363. @function2-llx please test it out.

**Comment by function2-llx** [comment]
@mauvilsa I've tested the fixed code, and it works fine now. Thanks!
